OrbitLinuxTracing/UprobesReturnAddressManager.h

```diff
@@ -59,6 +61,38 @@ class UprobesReturnAddressManager {
     }
   }
 
+  void PatchCallchain(pid_t tid, uint64_t* callchain, uint64_t nr,
+                      unwindstack::Maps* maps) {
+    if (!tid_uprobes_stacks_.contains(tid)) {
+      return;
+    }
+
+    auto& tid_uprobes_stack = tid_uprobes_stacks_.at(tid);
+    CHECK(!tid_uprobes_stack.empty());
+
+    // TODO(kuebler): What about tail-call optimization, where two uretprobes
+    //  hijacked an address at the same stack pointer.
+    auto uprobes_it = tid_uprobes_stack.rbegin();
+    for (uint64_t i = 0; i < nr; i++) {
+      uint64_t ip = callchain[i];
+      unwindstack::MapInfo* map_info = maps->Find(ip);
+
+      // Only patch Broken IPs
+      if (map_info == nullptr || map_info->name != "[uprobes]") {
+        continue;
+      }
+
+      const OpenUprobes& uprobes = *uprobes_it;
+      callchain[i] = uprobes.return_address;
+      uprobes_it++;
+
+      // There can't be more frames to patch as uprobes.
+      if (uprobes_it == tid_uprobes_stack.rend()) {
```

Comment:
What about checking this before the actual patching and printing an error if this is true?

Reply:
I was thinking about this first. The point here is that, if we have a callchain of lets say 100 frames, and there is just one uprobe, which is very much at the top of the chain, we can exit the loop quite early, as there can't be any other uprobe.

Reply:
Your point makes sense. Personally I think that for now it could be useful to have such errors while you make experiments, you can add a TODO explaining how that can be optimized later.
I would even add the opposite error, that is if after the loop `uprobes_it == tid_uprobes_stack.rend()` doesn't hold.
All this is especially since this code doesn't really have other ways to check that the correct frames are being patches.
In general and in other words, I would have some kind of error print if the number of uprobes frames in the callchain is different than the size of `tid_uprobes_stack`.